### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by colmazia**
Fixed: Refactor Environment in owasm

## Implementation details
- make branching operator cost more than the other operators
- change test to match the new cost function
- split VMLogic into BackendApi and Querier by level of host of each funtions

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [x] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [ ] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: packages/vm/src/calls.rs [hunk 1]
```diff
@@ -55,19 +57,25 @@ where
 #[cfg(test)]
 mod test {
     use crate::cache::CacheOptions;
+    use crate::vm::BackendApi;
```
Comment (taobun) [inline packages/vm/src/calls.rs]: super should import this

File: packages/vm/src/calls.rs [hunk 2]
```diff
@@ -1,23 +1,25 @@
-use crate::cache::Cache;
 use crate::error::Error;
 use crate::imports::create_import_object;
 use crate::store::make_store;
-use crate::vm::{Env, Environment};
+use crate::vm::{Environment, Querier};
+use crate::{cache::Cache, vm::BackendApi};
```
Comment (taobun) [inline packages/vm/src/calls.rs]: please fix

File: packages/vm/src/calls.rs [hunk 3]
```diff
@@ -55,16 +57,21 @@ where
 #[cfg(test)]
 mod test {
     use crate::cache::CacheOptions;
+    use crate::vm::BackendApi;
```
Comment (taobun) [inline packages/vm/src/calls.rs]: Remove

File: packages/vm/src/vm.rs [hunk 4]
```diff
@@ -6,7 +6,9 @@ use std::sync::{Arc, Mutex, RwLock};
 use wasmer::{Instance, Memory, WasmerEnv};
 use wasmer_middlewares::metering::{get_remaining_points, set_remaining_points, MeteringPoints};
 
-pub trait Env {
+pub trait BackendApi {}
```
Comment (taobun) [inline packages/vm/src/vm.rs]: Or should we remove this trait 🤣

File: packages/vm/src/vm.rs [hunk 5]
```diff
@@ -123,13 +100,22 @@ where
...
     where
-        C: FnOnce(&ContextData) -> R,
+        C: FnOnce(&ContextData<Q>) -> R,
     {
         let guard = self.data.as_ref().read().unwrap();
         let context_data = guard.borrow();
         callback(context_data)
     }
 
+    // fn with_context_data_mut<C, R>(&self, callback: C) -> R
```
Comment (taobun) [inline packages/vm/src/vm.rs]: Remove?

File: packages/vm/src/imports.rs [hunk 6]
```diff
@@ -163,31 +179,41 @@ where
...
-
-//         let y: Vec<u8> = get_from_mem(env, y_ptr, y_len)?;
-//         let pi: Vec<u8> = get_from_mem(env, pi_ptr, pi_len)?;
-//         let alpha: Vec<u8> = get_from_mem(env, alpha_ptr, alpha_len)?;
-//         Ok(ecvrf::ecvrf_verify(&y, &pi, &alpha) as u32)
-//     })
-// }
-
-pub fn create_import_object<E>(store: &Store, owasm_env: Environment<E>) -> ImportObject
+fn get_from_mem<A, Q>(env: &Environment<A, Q>, ptr: i64, len: i64) -> Result<Vec<u8>, Error>
```
Comment (taobun) [inline packages/vm/src/imports.rs]: Rename to `read_memory` or `read_region`

File: packages/vm/src/imports.rs [hunk 7]
```diff
@@ -127,33 +137,39 @@ where
             .iter()
```
Comment (taobun) [inline packages/vm/src/imports.rs]: Use read_memory function ?

**Review by taobun — CHANGES_REQUESTED** [review]

File: packages/vm/src/vm.rs (hunk 5 above)
Comment (colmazia) [reply packages/vm/src/vm.rs]: yes, but it is fundamental function that is likely to be used in the future so I think we better just commented it

File: packages/vm/src/imports.rs [hunk 8]
```diff
@@ -12,182 +12,193 @@ fn require_mem_range(max_range: usize, require_range: usize) -> Result<(), Error
...
 {
-    env.with_mut_vm(|vm| {
-        let span_size = vm.env.get_span_size();
+    env.with_querier_from_context(|querier| {
+        let span_size = querier.get_span_size();
 
         if len > span_size {
             return Err(Error::SpanTooSmallError);
         }
+        env.decrease_gas_left(1_750_000_000 + (len as u64) * 1_500_000)?;
```
Comment (taobun) [inline packages/vm/src/imports.rs]: Create gas calculate function when read memomy

File: packages/vm/src/imports.rs [hunk 9]
```diff
@@ -12,182 +12,193 @@ fn require_mem_range(max_range: usize, require_range: usize) -> Result<(), Error
...
 }
 
-fn do_gas<E>(env: &Environment<E>, _gas: u32) -> Result<(), Error>
+fn do_gas<Q>(env: &Environment<Q>, _gas: u32) -> Result<(), Error>
 where
-    E: Env + 'static,
+    Q: Querier + 'static,
 {
-    env.decrease_gas_left(12500000)
+    env.decrease_gas_left(750_000_000)?;
```
Comment (taobun) [inline packages/vm/src/imports.rs]: add const variable

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

File: packages/vm/src/cache.rs [hunk 10]
```diff
@@ -1,21 +1,14 @@
+use std::{
+    borrow::BorrowMut,
```
Comment (taobun) [inline packages/vm/src/cache.rs:2]: Unused import?

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

File: packages/vm/src/imports.rs [hunk 11]
```diff
@@ -12,182 +15,201 @@ fn require_mem_range(max_range: usize, require_range: usize) -> Result<(), Error
...
+    let memory = env.memory()?;
+    require_mem_range(memory.size().bytes().0, (ptr + data.len() as i64) as usize)?;
+    for (idx, byte) in data.iter().enumerate() {
+        memory.view()[ptr as usize + idx].set(*byte);
+    }
+    Ok(data.len() as i64)
+}
+
+fn calculate_read_memory_gas(len: i64) -> u64 {
+    1_000_000_000 + (len as u64) * 1_500_000
```
Comment (RogerKSI) [inline packages/vm/src/imports.rs]: as discussed, please prevent all overflow cases. otherwise, it will panic.

**Review by RogerKSI — CHANGES_REQUESTED** [review]

File: packages/vm/src/imports.rs [hunk 12]
```diff
@@ -12,182 +15,218 @@ fn require_mem_range(max_range: usize, require_range: usize) -> Result<(), Error
...
 
-fn do_gas<E>(env: &Environment<E>, _gas: u32) -> Result<(), Error>
+fn read_memory<Q>(env: &Environment<Q>, ptr: i64, len: i64) -> Result<Vec<u8>, Error>
 where
-    E: Env + 'static,
+    Q: Querier + 'static,
 {
-    env.decrease_gas_left(12500000)
+    let memory = env.memory()?;
+    require_mem_range(memory.size().bytes().0, (ptr.saturating_add(len)) as usize)?;
```
Comment (taobun) [inline packages/vm/src/imports.rs]: let last = (ptr.saturating_add(len)) as usize

File: packages/vm/src/imports.rs [hunk 13]
```diff
@@ -12,182 +15,218 @@ fn require_mem_range(max_range: usize, require_range: usize) -> Result<(), Error
...
 
-fn do_get_span_size<E>(env: &Environment<E>) -> i64
+fn write_memory<Q>(env: &Environment<Q>, ptr: i64, data: Vec<u8>) -> Result<i64, Error>
 where
-    E: Env + 'static,
+    Q: Querier + 'static,
 {
-    env.with_vm(|vm| vm.env.get_span_size())
+    let memory = env.memory()?;
+    require_mem_range(memory.size().bytes().0, (ptr.saturating_add(data.len() as i64)) as usize)?;
```
Comment (taobun) [inline packages/vm/src/imports.rs]: ditto

**Review by taobun — CHANGES_REQUESTED** [review]

File: packages/vm/src/imports.rs [hunk 14]
```diff
@@ -12,182 +15,218 @@ fn require_mem_range(max_range: usize, require_range: usize) -> Result<(), Error
...
 where
-    E: Env + 'static,
+    Q: Querier + 'static,
 {
-    env.with_mut_vm(|vm| -> Result<i64, Error> {
-        let span_size = vm.env.get_span_size();
-
-        let memory = env.memory()?;
-        require_mem_range(memory.size().bytes().0, (ptr + span_size) as usize)?;
+    env.with_querier_from_context(|querier| {
```
Comment (RogerKSI) [inline packages/vm/src/imports.rs:217]: why the old one needs a return type of callback but the new one doesn't?

File: packages/crypto/src/errors.rs [hunk 15]
```diff
@@ -0,0 +1,129 @@
+#[cfg(feature = "backtraces")]
```
Comment (RogerKSI) [inline packages/crypto/src/errors.rs]: think filename shouldn't have "s"

File: packages/vm/src/error.rs [hunk 16]
```diff
@@ -23,6 +23,8 @@ pub enum Error {
     MemoryOutOfBoundError = 14, // Out-of-bound memory access while executing the wasm script
```
Comment (RogerKSI) [inline packages/vm/src/error.rs:23]: should we also use thiserror like error in owasm-crypto?

File: packages/crypto/src/lib.rs [hunk 17]
```diff
@@ -2,3 +2,4 @@
 extern crate lazy_static;
 
 pub mod ecvrf;
+pub mod errors;
```
Comment (RogerKSI) [inline packages/crypto/src/lib.rs]: error.

**Review by RogerKSI — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

File: packages/vm/src/imports.rs [hunk 18]
```diff
@@ -12,198 +16,272 @@ fn require_mem_range(max_range: usize, require_range: usize) -> Result<(), Error
...
-    E: Env + 'static,
+    Q: Querier + 'static,
 {
-    env.with_vm(|vm| vm.env.get_span_size())
+    if ptr < 0 {
+        return Err(Error::MemoryOutOfBoundError);
+    }
+    let memory = env.memory()?;
+    require_mem_range(memory.size().bytes().0, safe_add(ptr, len)?)?;
+    Ok(memory.view()[safe_convert(ptr)?..safe_add(ptr, len)?]
```
Comment (RogerKSI) [inline packages/vm/src/imports.rs]: .get

File: packages/vm/src/imports.rs [hunk 19]
```diff
@@ -12,198 +16,272 @@ fn require_mem_range(max_range: usize, require_range: usize) -> Result<(), Error
...
 {
-    env.with_mut_vm(|vm| -> Result<i64, Error> {
-        let span_size = vm.env.get_span_size();
+    if ptr < 0 {
+        return Err(Error::MemoryOutOfBoundError);
+    }
+    let memory = env.memory()?;
+    require_mem_range(memory.size().bytes().0, safe_add(ptr, safe_convert(data.len())?)?)?;
+    for (idx, byte) in data.iter().enumerate() {
+        memory.view()[safe_add(ptr, safe_convert(idx)?)?].set(*byte);
```
Comment (RogerKSI) [inline packages/vm/src/imports.rs]: ditto

**Review by RogerKSI — APPROVED** [review]

**Pull request merged**